### PR TITLE
CI refactoring and extensions

### PR DESCRIPTION
**Pull request by petiaccja**
## Description

A general revamp of the CI on GitHub workflows:
- Cross-platform support: tests run on ubuntu and macos
- Separation of eve and gt4py:
  - Eve and gt4py have separate tox test environments
  - Eve and gt4py have separate YAML files and workflows
  - Gt4py tests no longer include eve tests
  - Sub-projects are only run when necessary
- Integration with codecov
  - Record code coverage
  - Quality gates on PRs

## Requirements

Before submitting this PR, please make sure:

- [ ] You have run the code checks, tests and documentation build successfully
- [ ] All fixes and all new functionality are tested and documentation is up to date
- [ ] You looked at the [review checklist](https://github.com/GridTools/gt4py/blob/master/REVIEW_CHECKLIST.txt)

Additionally, if this PR contains code authored by new contributors:

- [ ] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [ ] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -23,22 +21,24 @@ install_command = python -m pip install --no-cache-dir {opts} {packages}
 commands_pre =
     rm -Rf tests/_reports/coverage*
     atlas: pip install -i https://test.pypi.org/simple/ atlas4py
-commands =
-    pytest --cache-clear --cov -n auto -v -- {posargs}
-    pytest --doctest-modules --cov --cov-append -n auto -v src/
 commands_post =
-    coverage report --rcfile=setup.cfg
-    coverage html --rcfile=setup.cfg --show-contexts
```
Comment (egparedes) [inline tox.ini:31]: Why removing the `html` reports here? They are very useful when running `tox` locally during development

File: .github/workflows/tox.yml [hunk 2]
```diff
@@ -1,95 +0,0 @@
-name: Tox
```
Comment (egparedes) [inline .github/workflows/tox.yml:1]: Same comment as above about updating the link to the badge in the `README.md` file

File: .github/workflows/tox.yml [hunk 3]
```diff
@@ -1,95 +0,0 @@
...
-        tox_env_factor="${{ matrix.tox-env-factors }}"
-        pip install tox clang-format
-        tox -r -e py${pyversion_no_dot}-${tox_env_factor}
-    - name: Archive code coverage results
-      uses: actions/upload-artifact@v2
-      with:
-        name: code-coverage-report
-        path: tests/_reports/
-
-  tox-quickstart-guide:
```
Comment (egparedes) [inline .github/workflows/tox.yml:53]: It looks like the quickstart-guide related actions haven't been moved to another file but just removed. I would keep them.

**Review by egparedes — COMMENTED** [review]
I know this PR is still WIP so this is not really a review but just some quick feedback to think about. In general it looks like it's going in the right direction

File: tox.ini (hunk 1 above)
Comment (petiaccja) [reply tox.ini:31]: I don't use it locally myself, but I restored it for everyone who does.

File: .github/workflows/code_quality.yml [hunk 4]
```diff
@@ -1,12 +1,14 @@
-name: Quality Assurance
+name: Code Quality
```
[the remark this answers is not in the document]
Comment (petiaccja) [reply .github/workflows/code_quality.yml:1]: Done! I'll add the code coverage badge in another PR. I don't have access to the codecov settings page and it's a bit difficult without having codecov integration on the main branches.

File: .github/workflows/tox.yml (hunk 2 above)
Comment (petiaccja) [reply .github/workflows/tox.yml:1]: Updated as well.

File: .github/workflows/tox.yml (hunk 3 above)
Comment (petiaccja) [reply .github/workflows/tox.yml:53]: I suspect that was added to the CI on the main branch after I created the PR :) I added them to the current workflows.

**Review by DropD — APPROVED** [review]

**Pull request merged**